soco/zonegroupstate.py

```diff
@@ -0,0 +1,267 @@
+        self.visible_zones.clear()
+
+    def poll(self, soco):
+        """Poll using the provided SoCo instance and process the payload."""
+        # pylint: disable=protected-access
+        if time.monotonic() < self._cache_until:
+            self.total_requests += 1
+            _LOG.debug(
+                "Cache still active (GetZoneGroupState) during poll for %s",
+                soco.ip_address,
+            )
+            return
+
+        if soco._is_satellite:
+            # Satellites can return outdated information, use the parent
+            _LOG.debug(
+                "Poll request on satellite (%s), using parent (%s)",
+                soco.ip_address,
+                soco._satellite_parent.ip_address,
+            )
+            soco = soco._satellite_parent
+
+        zgs = soco.zoneGroupTopology.GetZoneGroupState()["ZoneGroupState"]
+        self.process_payload(zgs, "poll", soco.ip_address)
+
+    def process_payload(self, payload, source, source_ip):
+        """Update using the provided XML payload."""
+        self.total_requests += 1
+
+        def update_cache():
+            if source == "event":
+                timeout = EVENT_CACHE_TIMEOUT
+            else:
+                timeout = POLLING_CACHE_TIMEOUT
+            self._cache_until = time.monotonic() + timeout
+            _LOG.debug("Setting ZGS cache to %ss", timeout)
+
+        tree = normalize_zgs_xml(payload)
+        normalized_zgs = str(tree)
+        if normalized_zgs == self._last_zgs:
+            _LOG.debug(
+                "Duplicate ZGS received from %s (%s), ignoring", source_ip, source
+            )
+            update_cache()
+            return
+
+        self.processed_count += 1
+        _LOG.debug(
+            "Updating ZGS with %s payload from %s (%s/%s processed)",
+            source,
+            source_ip,
+            self.processed_count,
+            self.total_requests,
+        )
+
+        self.update_soco_instances(tree)
+        update_cache()
+        self._last_zgs = normalized_zgs
+
+    def parse_zone_group_member(self, member_element):
+        """Parse a ZoneGroupMember or Satellite element from Zone Group
+        State, create a SoCo instance for the member, set basic attributes
+        and return it."""
+        # pylint: disable=protected-access
+
+        # Create a SoCo instance for each member. Because SoCo
+        # instances are singletons, this is cheap if they have already
+        # been created, and useful if they haven't. We can then
+        # update various properties for that instance.
+        member_attribs = member_element.attrib
+        ip_addr = member_attribs["Location"].split("//")[1].split(":")[0]
+        zone = config.SOCO_CLASS(ip_addr)
+        for key, attrib in ZGS_ATTRIB_MAPPING.items():
+            setattr(zone, attrib, member_attribs.get(key))
+
+        for channel_map in list(
+            filter(None, [zone._channel_map, zone._ht_sat_chan_map])
```

Comment:
Ok so, programming style preferences aside, I think the `filter(None` trick is anything but easily parseable, especially compared to its list comprehension counterpart:
```python
       for channel_map in [
            map for map in [zone._channel_map, zone._ht_sat_chan_map] if map is not None
       ]:
```

Reply:
I personally find the `filter` method easier to follow, but I'm open to suggestions. Anyone else want to weigh in here? 😄

Reply:
The `filter()` approach looks fine to me. Appears to be quite standard usage: https://docs.python.org/3/library/functions.html#filter.

(I had to remind myself how `filter()` works, but that's just me.)

Reply:
> I personally find the filter method easier to follow

really, even with the None. I mean, first argument is supposed to be a callable that will determine whether we pass the object on or not. The fact that you can pass None is there instead seems a little special.

> (I had to remind myself how filter() works, but that's just me.)

:D my point exactly. Did you have to look stuff up to read the list comprehension? Or would you have to if the list comprehension had covered a `map?` or `reduce` :D ?

Oh man, I really regret that time that Guido let his guard down over summer and let all these things in ;p In any case, never mind me, I like the prospect of functional programming and all the little tools that come with it, I just find it inherently more implicit and "_explicit is better than implicit_". In any case, this is style _of semantics_ and I don't dictate that for SoCo, in general we tend to go with consensus on that. So if you are both ok with it, just ignore me :)